### PR TITLE
[CALCITE-6389] RexBuilder.removeCastFromLiteral does not preserve semantics for some…

### PR DESCRIPTION
**Pull request by mihaibudiu**
… types of literal

There are at least two other issues that I attempted to fix which uncover this bug, so I decided to make a separate PR for fixing it. This should make the other PRs simpler and less controversial.

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/sql/type/SqlTypeUtil.java [hunk 1]
```diff
@@ -1822,6 +1823,35 @@ public static RelDataType extractLastNFields(RelDataTypeFactory typeFactory,
...
+   * @param toType A DECIMAL type.
+   * @return whether the value is valid for the type
+   */
+  public static boolean canBeRepresentedExactly(@Nullable BigDecimal value, RelDataType toType) {
+    assert toType.getSqlTypeName() == SqlTypeName.DECIMAL;
+    if (value == null) {
+      return true;
+    }
+    value = value.stripTrailingZeros();
+    if (value.scale() < 0) {
```
Comment (snuyanzin) [inline core/src/main/java/org/apache/calcite/sql/type/SqlTypeUtil.java:1845]: it would be great also to have a test with negative scale

Comment (mihaibudiu) [reply core/src/main/java/org/apache/calcite/sql/type/SqlTypeUtil.java:1845]: Frankly negative scales are something I don't understand well.
Many SQL dialects do not accept negative scales, but Calcite allegedly does, although there aren't many tests about that.
Maybe we can file a separate issue about testing negative scales in Calcite, I expect it may uncover quite a few bugs.

Comment (mihaibudiu) [reply core/src/main/java/org/apache/calcite/sql/type/SqlTypeUtil.java:1845]: I also think that whether negative scales are accepted or not should be a property of the type system.

Comment (snuyanzin) [reply core/src/main/java/org/apache/calcite/sql/type/SqlTypeUtil.java:1845]: I haven't checked it however if you think that it is large area to work with, then separate issue could be probably also ok

Comment (mihaibudiu) [reply core/src/main/java/org/apache/calcite/sql/type/SqlTypeUtil.java:1845]: Yes, I think that many places in Calcite assume implicitly that decimal scales cannot be negative.

Comment (mihaibudiu) [reply core/src/main/java/org/apache/calcite/sql/type/SqlTypeUtil.java:1845]: what do we do about this PR in particular?

Comment (snuyanzin) [reply core/src/main/java/org/apache/calcite/sql/type/SqlTypeUtil.java:1845]: as you mentioned if there is a separate issue for negative scale is required, could you please file one in Calcite jira?
All the rest is ok to me, thanks

**Review by snuyanzin — APPROVED** [review]
